File: screen-stream/screen-stream-watch/ViewController.m
//
//  ViewController.m
//  screen-stream-watch
//
//  Created by xiang on 14/09/2017.
//  Copyright © 2017 dotEngine. All rights reserved.
//

#import "ViewController.h"

#import <DotEngine.h>
#import <DotStream.h>


static  NSString*  APP_KEY = @"45";
static  NSString*  APP_SECRET = @"dc5cabddba054ffe894ba79c2910866c";
static  NSString*  ROOM = @"screen_stream";


@interface ViewController ()<DotEngineDelegate,DotStreamDelegate>
{
    
    DotEngine* dotEngine;
    DotStream* localStream;
    DotStream* remoteStream;
    
    BOOL isStarted;
}


@property (weak, nonatomic) IBOutlet UIButton *startButton;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    dotEngine = [DotEngine sharedInstanceWithDelegate:self];
    localStream = [[DotStream alloc] initWithAudio:YES
                                             video:NO
                                          delegate:self];
    
    // Do any additional setup after loading the view, typically from a nib.
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (IBAction)start:(id)sender {
    
    self.startButton.enabled = false;
    
    [localStream setupLocalMedia];
    
    uint32_t randomNum = arc4random_uniform(10000);
    
    NSString* userId = [NSString stringWithFormat:@"watch%d",randomNum];
    
    [dotEngine generateTestTokenWithAppKey:APP_KEY
                                 appsecret:APP_SECRET
                                      room:ROOM
                                    userId:userId withBlock:^(NSString *token, NSError *error) {
                                        if (error == nil) {
                                            [dotEngine joinRoomWithToken:token];
                                        }
                                    }];
}



#pragma DotEngine Delegate


-(void)dotEngine:(DotEngine* _Nonnull) engine didJoined:(NSString* _Nonnull)peerId
{
    NSLog(@"didJoined ");
}

-(void)dotEngine:(DotEngine* _Nonnull) engine didLeave:(NSString* _Nonnull)peerId
{
    NSLog(@"didLeave");
}

-(void)dotEngine:(DotEngine* _Nonnull) engine  stateChange:(DotStatus)state
{
    if (state == DotStatusConnected) {
        [dotEngine addStream:localStream];
        self.startButton.hidden = true;
    }
}

-(void)dotEngine:(DotEngine* _Nonnull) engine didAddLocalStream:(DotStream* _Nonnull)stream
{
    NSLog(@"didAddLocalStream");
}

-(void)dotEngine:(DotEngine* _Nonnull) engine didRemoveLocalStream:(DotStream* _Nonnull)stream
{
    
}

-(void)dotEngine:(DotEngine* _Nonnull) engine didAddRemoteStream:(DotStream* _Nonnull)stream
{
    remoteStream = stream;
    
    remoteStream.view.frame = CGRectMake(0, 0, self.view.frame.size.width, self.view.frame.size.height);
    [remoteStream.view setScaleMode:DotVideoViewScaleModeFit];
    
    [self.view addSubview:remoteStream.view];
}

-(void)dotEngine:(DotEngine* _Nonnull) engine didRemoveRemoteStream:(DotStream* _Nonnull) stream
{
    if (remoteStream != nil && [remoteStream.streamId isEqualToString:stream.streamId]) {
        [remoteStream.view removeFromSuperview];
    }
}

-(void)dotEngine:(DotEngine* _Nonnull) engine didOccurError:(DotEngineErrorCode)errorCode
{
    
    NSLog(@"didOccurError  %ld", (long)errorCode);
}



#pragma DotStream delegate


-(void)stream:(DotStream* _Nullable)stream  didMutedVideo:(BOOL)muted
{
    
}

-(void)stream:(DotStream* _Nullable)stream  didMutedAudio:(BOOL)muted
{
    
}

-(void)stream:(DotStream* _Nullable)stream  didGotAudioLevel:(int)audioLevel
{
    
}



@end

File: screen-stream/screen-stream-watch-swift/ViewController.swift
//
//  ViewController.swift
//  screen-stream-watch-swift
//
//  Created by Gong Zhang on 2017/9/15.
//  Copyright © 2017年 dotEngine. All rights reserved.
//

import UIKit

let APP_KEY = "45"
let APP_SECRET = "dc5cabddba054ffe894ba79c2910866c"
let ROOM = "screen_stream"

class ViewController: UIViewController, DotEngineDelegate, DotStreamDelegate {
    
    @IBOutlet weak var startButton: UIButton!
    
    var dotEngine: DotEngine!
    var localStream: DotStream!
    var remoteStream: DotStream?
    
    var isStarted = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        dotEngine = DotEngine.sharedInstance(with: self)
        localStream = DotStream(audio: true, video: false, delegate: self)
    }
    
    @IBAction func startButtonAction(_ sender: Any) {
        startButton.isEnabled = false
        
        localStream.setupLocalMedia()
        
        let randomNum = arc4random_uniform(10000)
        let userId = "watch\(randomNum)"
        dotEngine.generateTestToken(withAppKey: APP_KEY, appsecret: APP_SECRET, room: ROOM, userId: userId) { [weak self] (token, err) in
            guard err == nil else {
                print(err!)
                return
            }
            
            guard let token = token else {
                print("token is nil")
                return
            }
            
            self?.dotEngine.joinRoom(withToken: token)
        }
    }
    
    // MARK: - DotEngine Delegates
    
    func dotEngine(_ engine: DotEngine, didJoined peerId: String) {
        print("didJoined")
    }
    
    func dotEngine(_ engine: DotEngine, didLeave peerId: String) {
        print("didLeave")
    }
    
    func dotEngine(_ engine: DotEngine, stateChange state: DotStatus) {
        if state == .connected {
            dotEngine.add(localStream)
            startButton.isHidden = true
        }
    }
    
    func dotEngine(_ engine: DotEngine, didAddLocalStream stream: DotStream) {
        print("didAddLocalStream")
    }
    
    func dotEngine(_ engine: DotEngine, didRemoveLocalStream stream: DotStream) {
    }
    
    func dotEngine(_ engine: DotEngine, didAddRemoteStream stream: DotStream) {
        remoteStream = stream
        stream.view?.frame = CGRect(x: 0, y: 0, width: view.bounds.size.width, height: view.bounds.size.height)
        stream.view?.scaleMode = .fit
        
        view.addSubview(stream.view!)
    }
    
    func dotEngine(_ engine: DotEngine, didRemoveRemoteStream stream: DotStream) {
        guard let remoteStream = remoteStream else {
            return
        }
        
        if remoteStream.streamId == stream.streamId {
            remoteStream.view?.removeFromSuperview()
            self.remoteStream = nil
        }
    }
    
    func dotEngine(_ engine: DotEngine, didOccurError errorCode: DotEngineErrorCode) {
        print("didOccurError \(errorCode)")
    }
    
    // MARK: - DotStream Delegates
    
    func stream(_ stream: DotStream?, didMutedVideo muted: Bool) {
    }
    
    func stream(_ stream: DotStream?, didMutedAudio muted: Bool) {
    }
    
    func stream(_ stream: DotStream?, didGotAudioLevel audioLevel: Int32) {
    }
    
}

